libs/gl-client-py/src/scheduler.rs

```diff
@@ -1,144 +1,264 @@
+use crate::credentials::{Credentials, PyCredentials};
 use crate::runtime::exec;
 use crate::Signer;
-use crate::{pb, pb::scheduler::scheduler_client::SchedulerClient};
-use anyhow::Result;
+use anyhow::{anyhow, Result};
 use gl_client::bitcoin::Network;
-use gl_client::tls::TlsConfig;
+use gl_client::credentials::RuneProvider;
+use gl_client::credentials::TlsConfigProvider;
+use gl_client::pb;
+use gl_client::scheduler;
 use prost::Message;
 use pyo3::exceptions::PyValueError;
 use pyo3::prelude::*;
-use tonic::transport::Channel;
 
-type Client = SchedulerClient<Channel>;
+pub enum UnifiedScheduler<T, R>
+where
+    T: TlsConfigProvider,
+    R: TlsConfigProvider + RuneProvider + Clone,
+{
+    Unauthenticated(scheduler::Scheduler<T>),
+    Authenticated(scheduler::Scheduler<R>),
+}
+
+impl<T, R> UnifiedScheduler<T, R>
+where
+    T: TlsConfigProvider,
+    R: TlsConfigProvider + RuneProvider + Clone,
+{
+    pub fn is_authenticated(&self) -> Result<()> {
+        if let Self::Authenticated(_) = self {
+            Ok(())
+        } else {
+            Err(anyhow!("scheduler is unauthenticated",))?
+        }
+    }
+
+    async fn register(
+        &self,
+        signer: &gl_client::signer::Signer,
+        invite_code: Option<String>,
+    ) -> Result<pb::scheduler::RegistrationResponse> {
+        match self {
+            UnifiedScheduler::Unauthenticated(u) => u.register(&signer, invite_code).await,
+            UnifiedScheduler::Authenticated(a) => a.register(&signer, invite_code).await,
+        }
+    }
+
+    async fn recover(
+        &self,
+        signer: &gl_client::signer::Signer,
+    ) -> Result<pb::scheduler::RecoveryResponse> {
+        match self {
+            UnifiedScheduler::Unauthenticated(u) => u.recover(&signer).await,
+            UnifiedScheduler::Authenticated(a) => a.recover(&signer).await,
+        }
+    }
+
+    async fn authenticate(&self, creds: R) -> Result<Self> {
```

Comment:
Could we constrain this to return an `Ok` when unauthenticated and an `Err` otherwise? I feel like preventing `Authenticated -> Authenticated` would help discourage instances of UnifiedSchedulers from being used as re-usable builders. Would it also be possible to pass `self` instead of `&self` to help enforce this?

Reply:
It would certainly be possible to consume the scheduler here but unfortunately we can not consume the scheduler in a pyo3 class, so this wouldn't help with the public api. Let me just add an `Err` if the scheduler is already authenticated. I guess we should do the same in the rust library then?